tests/DfE.FindInformationAcademiesTrusts.Data.AcademiesDb.UnitTests/DataSourceProviderTests.cs

```diff
@@ -0,0 +1,154 @@
+            null, UpdateFrequency.Daily));
+    }
+
+    [Fact]
+    public async Task GetMstrUpdated_WhenEntryExists_ShouldReturnDataSource()
+    {
+        var result = await _sut.GetMstrUpdated();
+
+        result.Should().NotBeNull();
+        result.Should()
+            .BeEquivalentTo(new DataSource(Source.Mstr, TestStartTime.AddDays(-1),
+                UpdateFrequency.Daily));
+    }
+
+    [Fact]
+    public async Task GetMstrUpdated_WhenInvalidEntryExists_ShouldReturnDataSource_WithNullDate()
+    {
+        _mockAcademiesDbContext.SetupInvalidMockDbContextOpsApplicationEvents(TestStartTime);
+        var result = await _sut.GetMstrUpdated();
+
+        result.Should().NotBeNull();
+        result.Should().BeEquivalentTo(new DataSource(Source.Mstr,
+            null, UpdateFrequency.Daily));
+    }
+
+    [Fact]
+    public async Task GetCdmUpdated_WhenNoEntryExists_ShouldReturnDataSource_WithNullDate()
+    {
+        _mockAcademiesDbContext.SetupEmptyMockDbContextOpsApplicationEvents();
+        var result = await _sut.GetCdmUpdated();
+
+        result.Should().NotBeNull();
+        result.Should().BeEquivalentTo(new DataSource(Source.Cdm,
+            null, UpdateFrequency.Daily));
+    }
+
+    [Fact]
+    public async Task GetCdmUpdated_WhenEntryExists_ShouldReturnDataSource()
+    {
+        var result = await _sut.GetCdmUpdated();
+
+        result.Should().NotBeNull();
+        result.Should()
+            .BeEquivalentTo(new DataSource(Source.Cdm, TestStartTime.AddDays(-1),
+                UpdateFrequency.Daily));
+    }
+
+    [Fact]
+    public async Task GetCdmUpdated_WhenInvalidEntryExists_ShouldReturnDataSource_WithNullDate()
+    {
+        _mockAcademiesDbContext.SetupInvalidMockDbContextOpsApplicationEvents(TestStartTime);
+        var result = await _sut.GetCdmUpdated();
+
+        result.Should().NotBeNull();
+        result.Should()
+            .BeEquivalentTo(new DataSource(Source.Cdm,
+                null, UpdateFrequency.Daily));
+    }
+
+    [Fact]
+    public async Task GetMisEstablishmentsUpdated_WhenNoEntryExists_ShouldReturnDataSource_WithNullDate()
+    {
+        _mockAcademiesDbContext.SetupEmptyMockDbContextOpsApplicationSettings();
+        var result = await _sut.GetMisEstablishmentsUpdated();
+
+        result.Should().NotBeNull();
+        result.Should()
+            .BeEquivalentTo(new DataSource(Source.Mis, null, UpdateFrequency.Monthly));
+    }
+
+    [Fact]
+    public async Task GetMisEstablishmentsUpdated_WhenEntryExists_ShouldReturnDataSource()
+    {
+        var result = await _sut.GetMisEstablishmentsUpdated();
+
+        result.Should().NotBeNull();
+        result.Should()
+            .BeEquivalentTo(new DataSource(Source.Mis,
+                TestStartTime.AddDays(-1), UpdateFrequency.Monthly));
+    }
+
+    [Fact]
+    public async Task GetMisEstablishmentsUpdated_WhenInvalidEntryExists_ShouldReturnDataSource_WithNullDate()
```

Comment:
We should also have assertions to ensure that we're logging the right thing on invalid/missing entries